# MISM-6205 
# Project 2: Data Profiling and Preparation
## Group 11
### Sarang Padhye
### Joseph Akkara
### Sushil Goud Balakrishna

# 1. Conduct Data Access and Profiling.

## 1. Read the survey dataset into a pandas DataFrame. Importing the needed packages

In [1]:
import pandas as pd

In [2]:
print("This is the initial dataset:")
df = pd.read_csv ('Passengers_Satisfaction_Survey.csv', low_memory=False)
df

This is the initial dataset:


Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
1         Male  Disloyal Customer  20.0  Business travel       Eco   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0                  32.0                    3.0   
1                   0.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
1                                     3.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
1                   2                    10.0                4          2.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes Arrival Delay in Minutes  Satisfaction  
0                             14.0                        0  Dissatisfied  
1                              0.0                        0  dissatisfied  
2                              NaN                      NaN       neutral  
3                              0.0                        6  Dissatisfied  
4                              NaN                      NaN       neutral  
...                            ...                      ...           ...  
104905                         8.0                        0     satisfied  
104906                        54.0                       40     satisfied  
104907                         0.0                        2     satisfied  
104908                         0.0      

In [3]:
print("The dimensions of the initial dataset is")
print(df.shape)

The dimensions of the initial dataset is
(104910, 16)


## 2. Create a subset of the DataFrame to include all passenger satisfaction data about only the Business Class. Name the created DataFrame subset “df_Business”.

In [4]:
# Creating new dataframe df_Business
print("The new dataframe with only Business Class details is")
df_Business = df[(df["Class"] == "Business")] 
df_Business

The new dataframe with only Business Class details is


Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
5       Female     Loyal Customer  57.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0                  32.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
5                   0.0                    3.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
5                                     4.0                      4   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
5                   5                     2.0                3          2.0   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes Arrival Delay in Minutes  Satisfaction  
0                             14.0                        0  Dissatisfied  
2                              NaN                      NaN       neutral  
3                              0.0                        6  Dissatisfied  
4                              NaN                      NaN       neutral  
5                              0.0                        0  Dissatisfied  
...                            ...                      ...           ...  
104905                         8.0                        0     satisfied  
104906                        54.0                       40     satisfied  
104907                         0.0                        2     satisfied  
104908                         0.0      

## 3. Collect information about the “df_Business” dataset and its validity. Report the following: 

### 3.1 The number of observations (or rows) and the number of variables (or columns) in the dataset.

In [5]:
print("The dimensions of the new dataframe are:\n ")
df_Business.shape

The dimensions of the new dataframe are:
 


(50000, 16)

### 3.2 The name and data type of each data column.

In [6]:
print("Below are the columns and their data types", '\n')
df_Business.info()

Below are the columns and their data types 

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handl

### 3.3 The unique values of each column.

In [7]:
print("Below are the unique values from all columns in the df_Business dtaframe")
cols = df_Business.columns

# for each column
for col in cols:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print(unique, '\n\n-----------------------------------------------\n')

Below are the unique values from all columns in the df_Business dtaframe
Gender
['Male' 'Female' 'F' 'M' nan 'Fe male'] 

-----------------------------------------------

Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 

-----------------------------------------------

Age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 

-----------------------------------------------

Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 

-----------------------------------------------

Class
['Business'] 

-----------------------------------------------

Flight Distance
[3.200e+01 0.000e+00 5.600e+01 ... 4.983e+03 2.820e+06 9.000

### 3.4 The number of missing values of each column.

In [8]:
print("Below are the number of missing values from all columns \n")
df_Business.isnull().sum()

Below are the number of missing values from all columns 



Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

### 3.5 Summary statistics for each data column.

In [9]:
print("\n Below are the statistics for all columns")
df_Business.describe(include='all')


 Below are the statistics for all columns


Gender   Customer Type           Age   Type of Travel     Class  \
count    48239           48239  47373.000000            48239     50000   
unique       5               4           NaN                3         1   
top     Female  Loyal Customer           NaN  Business travel  Business   
freq     23000           40770           NaN            44063     50000   
mean       NaN             NaN     41.853186              NaN       NaN   
std        NaN             NaN     13.730683              NaN       NaN   
min        NaN             NaN      0.000000              NaN       NaN   
25%        NaN             NaN     34.000000              NaN       NaN   
50%        NaN             NaN     43.000000              NaN       NaN   
75%        NaN             NaN     52.000000              NaN       NaN   
max        NaN             NaN    185.000000              NaN       NaN   

        Flight Distance  Inflight wifi service  \
count      4.660400e+04           48239.000000   
unique              NaN                    NaN   
top                 NaN                    NaN   
freq                NaN                    NaN   
mean       1.934729e+03               2.774519   
std        4.369344e+04               1.423917   
min       -5.421000e+03               0.000000   
25%        5.940000e+02               2.000000   
50%        1.597000e+03               3.000000   
75%        2.565000e+03               4.000000   
max        9.000000e+06               5.000000   

        Departure/Arrival time convenient Ease of Online booking  \
count                        48239.000000                  42705   
unique                                NaN                      7   
top                                   NaN                      3   
freq                                  NaN                   8635   
mean                             2.910695                    NaN   
std                              1.501711                    NaN   
min                              0.000000                    NaN   
25%                              2.000000                    NaN   
50%                              3.000000                    NaN   
75%                              4.000000                    NaN   
max                              5.000000                    NaN   

       Food and drink  Inflight entertainment Baggage handling   Cleanliness  \
count           42705            43649.000000            43649  43649.000000   
unique              7                     NaN                6           NaN   
top                 4                     NaN                4           NaN   
freq            10890                     NaN            17056           NaN   
mean              NaN                3.835391              NaN      3.737520   
std               NaN                1.233508              NaN      0.987664   
min               NaN                0.000000              NaN      0.000000   
25%               NaN                3.000000              NaN      3.000000   
50%               NaN                4.000000              NaN      4.000000   
75%               NaN                5.000000              NaN      5.000000   
max               NaN               10.000000              NaN      5.000000   

        Departure Delay in Minutes Arrival Delay in Minutes Satisfaction  
count                 43649.000000                    42098        47840  
unique                         NaN                      363            6  
top                            NaN                        0    satisfied  
freq                           NaN                    24229        30398  
mean                     13.595363                      NaN          NaN  
std                      36.595048                      NaN          NaN  
min                      -2.000000                      NaN          NaN  
25%                       0.000000                      NaN          NaN  
50%                       0.000000                   

### 3.6 Frequency distribution (list the number of times each unique value appears) for each data column.

In [10]:
print("\nNumber of times each unique value appears in all columns \n")
cols = df_Business.columns
for col in cols:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n------------------------------------------\n\n')


Number of times each unique value appears in all columns 

Gender
Gender
Female     23000
Male       22830
F           1361
M           1047
Fe male        1
Name: count, dtype: int64 

------------------------------------------


Customer Type
Customer Type
Loyal Customer       40770
Disloyal Customer     6636
disloyal               831
Loyal                    2
Name: count, dtype: int64 

------------------------------------------


Age
Age
39.0     1935
40.0     1677
41.0     1533
42.0     1514
44.0     1506
         ... 
73.0       22
78.0       17
85.0       14
180.0       1
185.0       1
Name: count, Length: 74, dtype: int64 

------------------------------------------


Type of Travel
Type of Travel
Business travel    44063
Business            2122
Personal Travel     2054
Name: count, dtype: int64 

------------------------------------------


Class
Class
Business    50000
Name: count, dtype: int64 

------------------------------------------


Flight Distance
Flight Distance

### 3.7 The number of fully duplicated data rows.

In [11]:
print("Boolean output for duplicate data rows\n")
df_Business.duplicated()

Boolean output for duplicate data rows



0         False
2         False
3         False
4         False
5         False
          ...  
104905    False
104906    False
104907    False
104908    False
104909    False
Length: 50000, dtype: bool

In [12]:
print("\nThe number of duplicated data rows are:\n")
df_Business.duplicated().sum()


The number of duplicated data rows are:



3064

In [13]:
# Selecting duplicate based on all columns
duplicate = df_Business[df_Business.duplicated()]
  
print("The Duplicate Data Rows are:")
duplicate

The Duplicate Data Rows are:


Gender      Customer Type   Age   Type of Travel     Class  \
153          F  Disloyal Customer  25.0         Business  Business   
349        NaN                NaN   NaN              NaN  Business   
369        NaN                NaN   NaN              NaN  Business   
705        NaN                NaN   NaN              NaN  Business   
1159    Female     Loyal Customer  49.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104737  Female     Loyal Customer  37.0  Business travel  Business   
104768     NaN                NaN   NaN              NaN  Business   
104789    Male     Loyal Customer  24.0         Business  Business   
104825       M     Loyal Customer  30.0  Business travel  Business   
104840       M     Loyal Customer  30.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
153                67.0                    1.0   
349                 NaN                    NaN   
369                 NaN                    NaN   
705                 NaN                    NaN   
1159              101.0                    1.0   
...                 ...                    ...   
104737           3987.0                    3.0   
104768              NaN                    NaN   
104789           3993.0                    2.0   
104825              NaN                    3.0   
104840              NaN                    4.0   

        Departure/Arrival time convenient Ease of Online booking  \
153                                   0.0                      1   
349                                   NaN                    NaN   
369                                   NaN                    NaN   
705                                   NaN                    NaN   
1159                                  4.0                      4   
...                                   ...                    ...   
104737                                1.0                      3   
104768                                NaN                    NaN   
104789                                5.0                      5   
104825                                3.0                    NaN   
104840                                4.0                    NaN   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
153                 3                     3.0                4          3.0   
349               NaN                     NaN              NaN          NaN   
369               NaN                     NaN              NaN          NaN   
705               NaN                     NaN              NaN          NaN   
1159                3                     2.0                1          3.0   
...               ...                     ...              ...          ...   
104737              5                     5.0                5          5.0   
104768            NaN                     NaN              NaN          NaN   
104789              2                     2.0                A          2.0   
104825            NaN                     4.0                1          4.0   
104840            NaN                     4.0                4          4.0   

        Departure Delay in Minutes Arrival Delay in Minutes  Satisfaction  
153                            0.0                        0           NaN  
349                            NaN                      NaN  dissatisfied  
369                            NaN                      NaN  dissatisfied  
705                            NaN                      NaN  dissatisfied  
1159                          11.0                       14  Dissatisfied  
...                            ...                      ...           ...  
104737                        10.0                       23     satisfied  
104768                         NaN                      NaN  dissatisfied  
104789                        23.0                      -45  dissatisfied  
104825                        10.0      

In [14]:
duplicate.shape

(3064, 16)


## Summary 
### List of the issues that the data contains

1. Initial Customer Satisfaction Dataset has 104910 rows and 16 columns. 
2. New dataframe with only Business Class has 50000 rows.
2. The new dataframe has a column i.e. column 14 Departure delay in Minutes, which has mixed datatypes.
3. The new dataframe has 3064 duplicate rows.
4. These columns contain float and categorical values.
5. "Age" should be in integer format instead of float.
6. "Inflight Wi-Fi Service", "Departure/Arrival Time Convenient","Inflight entertainment" and "Cleanliness" should be an integer. are in float format.
7. "Ease of Online booking", "Food and drink", "Baggage handling" are in object format.
8. There are format error in column values of 'Gender', 'Customer Type', 'Type of Travel' and 'Satisfaction'.
9. The "Age" column has entries as 180 and 185 which are not the thr right inputs. So they can be removed as they values appear to be outliers.
10. The "Age" column contains numeric values, but there is an entry with "0," which isnt not the right input for age. So those entries need to be removed.
11. The column "Class" has single value and does not help extract any insights. Thus, we will eliminate it.
12. The columns "Ease of Online booking", "Food and Drink" and "Baggage Handling" have values which are non-numeric.
13. The columns "Flight Distance", "Departure Delay in Minutes", "Arrival Delay in Minutes" have negative values and improper data, namely "9.." 
14. The column "Flight Distance" has 2 entries for distances which are not meaningful with values of 2820000 and 9000000.

# 2. Conduct Data Cleansing:

### 2.1: Drop the Class column from the created Business Class passenger satisfaction survey (“df_Business”).

In [15]:
df_Business = df_Business.drop('Class', axis=1)

In [16]:
df_Business

Gender      Customer Type   Age   Type of Travel  Flight Distance  \
0         Male     Loyal Customer  69.0  Personal Travel             32.0   
2         Male  Disloyal Customer  41.0  Business travel              0.0   
3         Male     Loyal Customer  52.0  Business travel              0.0   
4       Female  Disloyal Customer  21.0  Business travel              0.0   
5       Female     Loyal Customer  57.0  Business travel              0.0   
...        ...                ...   ...              ...              ...   
104905    Male     Loyal Customer  45.0  Business travel           3305.0   
104906  Female     Loyal Customer  35.0  Business travel            337.0   
104907    Male     Loyal Customer  47.0  Business travel           3496.0   
104908    Male     Loyal Customer  42.0  Business travel           3605.0   
104909  Female           disloyal   0.0  Business travel            862.0   

        Inflight wifi service  Departure/Arrival time convenient  \
0                         3.0                                2.0   
2                         1.0                                1.0   
3                         2.0                                4.0   
4                         4.0                                0.0   
5                         3.0                                4.0   
...                       ...                                ...   
104905                    2.0                                2.0   
104906                    5.0                                5.0   
104907                    2.0                                2.0   
104908                    4.0                                4.0   
104909                    5.0                                5.0   

       Ease of Online booking Food and drink  Inflight entertainment  \
0                           3              5                     1.0   
2                         NaN            NaN                     NaN   
3                           4              1                     2.0   
4                         NaN            NaN                     NaN   
5                           4              5                     2.0   
...                       ...            ...                     ...   
104905                      2              3                     4.0   
104906                      5              2                     4.0   
104907                      2              2                     5.0   
104908                      4              3                     4.0   
104909                      5              5                     5.0   

       Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                     1          3.0                        14.0   
2                   NaN          NaN                         NaN   
3                     2          4.0                         0.0   
4                   NaN          NaN                         NaN   
5                     3          2.0                         0.0   
...                 ...          ...                         ...   
104905                4          3.0                         8.0   
104906                4          5.0                        54.0   
104907                5          4.0                         0.0   
104908                4          3.0                         0.0   
104909                1          5.0                         9.0   

       Arrival Delay in Minutes  Satisfaction  
0                             0  Dissatisfied  
2                           NaN       neutral  
3                             6  Dissatisfied  
4                           NaN       neutral  
5                             0  Dissatisfied  
...                         ...           ...  
104905                        0     satisfied  
104906                       40     satisfied  
104907                        2     satisfied  
104908                        0     satisfied  
104909                        5       neutral  

[50000 

In [17]:
df_Business = df_Business.reset_index(drop=True)
df_Business.shape

(50000, 15)

### 2.2: Rename “df_Business” columns by making all the columns’ names in lowercase, and if the column name has whitespace, replace it with an underscore. (For example: “Customer Type” will be renamed to “customer_type”).

In [18]:
df_Business.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)
df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'flight_distance',
       'inflight_wifi_service', 'departure/arrival_time_convenient',
       'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment',
       'baggage_handling', 'cleanliness', 'departure_delay_in_minutes',
       'arrival_delay_in_minutes', 'satisfaction'],
      dtype='object')

In [19]:
print("Upated Column Names: \n")
for column in df_Business.columns:
    print(column)

Upated Column Names: 

gender
customer_type
age
type_of_travel
flight_distance
inflight_wifi_service
departure/arrival_time_convenient
ease_of_online_booking
food_and_drink
inflight_entertainment
baggage_handling
cleanliness
departure_delay_in_minutes
arrival_delay_in_minutes
satisfaction


### 2.3: Modify improperly formatted data and handle bad data.

In [20]:
col = df_Business.columns

for col in df_Business.columns:
    print(col,'column has', len(df_Business[col].unique()),'unique values')
    print(df_Business[col].unique())
    print('Frequency distribution of', col, 'is as follows:')
    print(df_Business[col].value_counts())
    print('-------------------------------------------------')

gender column has 6 unique values
['Male' 'Female' 'F' 'M' nan 'Fe male']
Frequency distribution of gender is as follows:
gender
Female     23000
Male       22830
F           1361
M           1047
Fe male        1
Name: count, dtype: int64
-------------------------------------------------
customer_type column has 5 unique values
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal']
Frequency distribution of customer_type is as follows:
customer_type
Loyal Customer       40770
Disloyal Customer     6636
disloyal               831
Loyal                    2
Name: count, dtype: int64
-------------------------------------------------
age column has 75 unique values
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  1

#### inflight_wifi_service, departure/arrival_time_convenient, cleanliness. These columns have no improperly formatted data, hence have not been considered in below code. 

In [21]:
condition = (df_Business['age'] >=180)
df_Business = df_Business.drop(df_Business[condition].index)

In [22]:
condition2 = (df_Business['age'] == 0)
df_Business = df_Business.drop(df_Business[condition2].index)

In [23]:
mode_value = df_Business['ease_of_online_booking'].dropna().mode()[0]
#mode_value = df_Business['ease_of_online_booking'].mode()[0]
mode_value
df_Business['ease_of_online_booking']=df_Business['ease_of_online_booking'].replace(['Unknown'],[mode_value])

In [24]:
df_Business['food_and_drink'] = pd.to_numeric(df_Business['food_and_drink'], errors='coerce')
mean_value = round(df_Business['food_and_drink'].mean())
mean_value_str = str(mean_value)
df_Business['food_and_drink']=df_Business['food_and_drink'].replace(['Great'],[mean_value_str])

In [25]:
df_Business['gender']= df_Business['gender'].replace(['F','Fe male','M'],['Female','Female','Male'])
df_Business['customer_type']= df_Business['customer_type'].replace(['Loyal','disloyal'],['Loyal Customer','Disloyal Customer'])
df_Business['type_of_travel']=df_Business['type_of_travel'].replace(['Business','Business travel'],['Business Travel','Business Travel'])
df_Business['inflight_entertainment']=df_Business['inflight_entertainment'].replace([10.],[5.])
df_Business['baggage_handling']=df_Business['baggage_handling'].replace(['A'],['5'])
df_Business['arrival_delay_in_minutes']=df_Business['arrival_delay_in_minutes'].replace(['Three'],['3'])
df_Business['satisfaction'] = df_Business['satisfaction'].replace(['neutral','dissatisfied','satisfied','0','1'],['Neutral','Dissatisfied','Satisfied','Dissatisfied','Satisfied'])

### 2.4 Handle the missing values (You should have zero missing values in your cleaned dataset). rating mode, integer float mean

In [26]:
df_Business['gender'].fillna('Unknown',inplace=True)
df_Business['customer_type'].fillna('Unknown',inplace=True)
df_Business['type_of_travel'].fillna('Unknown',inplace=True)
df_Business['satisfaction'].fillna('Unknown',inplace=True)
mean_value = round(df_Business['age'].mean())
mean_value
df_Business['age'].fillna(mean_value,inplace=True)
df_Business['inflight_wifi_service'].fillna(df_Business['inflight_wifi_service'].mode()[0],inplace=True)
df_Business['departure/arrival_time_convenient'].fillna(df_Business['departure/arrival_time_convenient'].mode()[0],inplace=True)
df_Business['ease_of_online_booking'].fillna(df_Business['ease_of_online_booking'].mode()[0],inplace=True)
df_Business['food_and_drink'].fillna(df_Business['food_and_drink'].mode()[0],inplace=True)
df_Business['inflight_entertainment'].fillna(df_Business['inflight_entertainment'].mode()[0],inplace=True)
df_Business['baggage_handling'].fillna(df_Business['baggage_handling'].mode()[0],inplace=True)
df_Business['cleanliness'].fillna(df_Business['cleanliness'].mode()[0],inplace=True)
df_Business['departure_delay_in_minutes'].fillna(df_Business['departure_delay_in_minutes'].mode()[0],inplace=True)
df_Business['flight_distance'].fillna(df_Business['flight_distance'].mode()[0],inplace=True)
df_Business['arrival_delay_in_minutes'].fillna(df_Business['arrival_delay_in_minutes'].mode()[0],inplace=True)

In [27]:
# In the flight_distance column, there are total 8 rows which have values for the distance of flight 
# as 0, -5421, 2820000, 9000000. So we have dropped these rows using 2 conditions.


condition1 = (df_Business['flight_distance'] > 100000)
df_Business = df_Business.drop(df_Business[condition1].index)

In [28]:
condition2 = (df_Business['flight_distance'] <= 0)
df_Business = df_Business.drop(df_Business[condition2].index)

In [29]:
condition3 = (df_Business['departure_delay_in_minutes'] < 0)
df_Business = df_Business.drop(df_Business[condition3].index)

In [30]:
condition4 = (df_Business['arrival_delay_in_minutes'] == '-45')
df_Business = df_Business.drop(df_Business[condition4].index)

In [31]:
condition5 = (df_Business['arrival_delay_in_minutes'] == '9..')
df_Business = df_Business.drop(df_Business[condition5].index)

### 2.5 Remove the fully duplicated observations.

In [32]:
df_Business = df_Business.drop_duplicates()

### 2.6 Fix the columns datatypes.

In [34]:
df_Business = df_Business.convert_dtypes()

# 3. Export the cleaned dataset to CSV file

In [35]:
df_Business.to_csv('Project_2_Group11.csv', index=False)

# 4. Compare the Data Characteristics Before and After Data Cleansing.

### At the end of the Jupyter Notebook, provide a brief write-up annotation to compare the characteristics of each column before and after data cleansing (this includes, the column name, datatype, unique values, and the number of missing values).

# The below is our comparison of the df_Business dataset before and after data cleansing


### 1. We updated the column names by changing them to lowercase & changed the separator from " " to "_".

    Column Names before data cleansing           Column Names after data cleansing.
    ---------------------------------           ---------------------------------
    Gender                      	            gender 
    Customer Type 	                            customer_type 
    Age 	                                    age 
    Type of Travel 	                            type_of_travel 
    Class 	
    Flight Distance                             flight_distance 
    Inflight wifi service 	                    inflight_wifi_service 
    Departure/Arrival time convenient           departure/arrival_time_convenient 
    Ease of Online booking 	                    ease_of_online_booking 
    Food and drink 	                            food_and_drink 
    Inflight entertainment 	                    inflight_entertainment 
    Baggage handling                            baggage_handling 
    Cleanliness 	                            cleanliness 
    Departure Delay in Minutes                  departure_delay_in_minutes 
    Arrival Delay in Minutes                    arrival_delay_in_minutes 
    Satisfaction                                satisfaction 

### 2. The data types for the columns before and after data cleansing are as below:

Before dara cleansing, some columns were initialized with the misappropriate data types. After pre-processing and cleansing the dataset, the NaN, Null values, wrong inputs were eliminated and the columns now have updated data types.

    Data Types before data cleansing                      Data Types after data cleansing.
    ---------------------------------                     ---------------------------------

    Gender                               object            gender                               string  
    Customer Type                        object            customer_type                        string  
    Age                                  float64           age                                  Int64  
    Type of Travel                       object            type_of_travel                       string 
    Class                                object            flight_distance                      Int64  
    Flight Distance                      float64           inflight_wifi_service                Int64  
    Inflight wifi service                float64           departure/arrival_time_convenient    Int64  
    Departure/Arrival time convenient    float64           ease_of_online_booking               string  
    Ease of Online booking               object            food_and_drink                       Int64  
    Food and drink                       object            inflight_entertainment               Int64  
    Inflight entertainment               float64           baggage_handling                     string  
    Baggage handling                     object            cleanliness                          Int64  
    Cleanliness                          float64           departure_delay_in_minutes           Int64  
    Departure Delay in Minutes           float64           arrival_delay_in_minutes             string  
    Arrival Delay in Minutes             object            satisfaction                         string  
    Satisfaction                         object  


#### Number of Observations and Variables.

Before Pre-Processing: The dataset contained 104911 observations (rows) and 16 variables (columns).
After Pre-processing: The df_Business dataframe now contains 50000 observations and 15 variables, as the 'Class' column was removed.

#### Unique Values

Before Data Cleaning: We detected columns with a significant number of distinct values, suggesting a wide range of inputs or potentially unwanted data. Before data cleansing, the unique values also consisted of error values or improper data like "F", "M", "Fe male", "Loyal", "disloyal", non-numeric values like "Unknown", "Great", "A" in numeric columns

After Data Cleaning: Some columns now exhibit a decreased count of unique values owing the standardizing of dataframe, signifying a more refined dataset. These were not only unique values but also improper data.


#### Missing Data

Before Data Cleaning: Various columns had "64884" number of missing values, potentially impacting the quality of any subsequent data processing.

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160


After Data Cleaning: All missing values have been resolved, either through data imputation by mode or mean values or by removing rows or columns, leading to a dataset with no remaining gaps.


#### Duplicate Rows

Before Cleansing: There were 3064 duplicate rows in the initial dataset which could skew our analysis.
After Cleansing: All duplicate rows have been removed, ensuring the uniqueness of each row.


To Summarize, the data cleansing process has improved the dataset's overall structure and quality. It makes the dataset more suitable ti gain insights from it. Dropping the unnecessary column, correction of data types, and handling of missing values have optimised the dataset. Additionally, the removal of duplicates and standardizing of unique values has increased the consistency of the dataset.